Test output formatting for strings, memory and bignums

paulidale opened the pull request:
- [x] tests are added or updated

Addition of user callable routines to output memory block, strings and bignums in a standard format that is equivalent to the test failure output of these (sans the diff).

The relevant test cases that did their own thing, now call the library routines.

The output code has been moved to its own file rather than cluttering up the tests source file.

levitte commented on test/exptest.c:
```diff
@@ -39,10 +32,9 @@ static int a_is_zero_mod_one(const char *method, const BIGNUM *r,
-        BN_print_var(r);
+        TEST_error("%s failed: a ** 0 mod 1 = r (should be 0)", method);
+        test_output_bignum("a", a);
+        test_output_bignum("r", r);
```
Any reason you didn't keep the `BN_print_var` calls here, specifically?

paulidale replied:
Only because I replaced the whole function before fixing the macro.
I'll switch it back to the macro.

paulidale commented:
Just fixed the missing file -- I do this one all too often :(

paulidale commented:
Fixed.

levitte commented on test/testutil/format_output.c:
```diff
@@ -0,0 +1,505 @@
+        test_printf_stderr("%*s# % 4s %c NULL\n", indent, "", "", c);
+    else
+        test_printf_stderr("%*s# % 4u:%c ''\n", indent, "", 0u, c);
+}
```
Not sure why this function, specifically, should be passed `indent` when all other helper functions of this sort simply use `subtest_level()`.

paulidale replied:
I don't remember why this one came about.  It isn't new code.  Still, I'll change it.

paulidale commented:
> Not that I'm fond of # signs all over, but looking at ectest output...

These are for the test framework output I believe @levitte ?


>      y^2 = x^3 + a*x + b  (mod 0x
> 
> Should second line be #-ed? Not to mention that it's unreadable by itself. I mean original output was "(mod 0x17)", and what's "(mod 0x"? Then is there a way to group them better?

Yes it should, I hadn't got to replacing printf's with test functions so the newline confused things.  I've added a new call TEST_note which acts like TEST_info but without the file/line information.  I've modified `ectest` to use this.


> And question is if one can (if appropriate?) get rid of empty lines between related bignums, i.e.

I've done this.  User requested output omits the lines.  Test failure output has them.


> On side note one can probably argue that it would be appropriate to see if bignum is not that big (not wider than 64 bits) and make printout friendlier by placing everything in one line, e.g.

This is a good suggestion.  I'll have a think about it.

paulidale commented:
I've added the short form output for 64 bit or less bignums.

dot-asm commented:
>> Not that I'm fond of # signs all over, but looking at ectest output...

> These are for the test framework output I believe @levitte ?

Remark was rather about mixture of #-ed lines and non-#-ed lines than # signs per se. Yes, I care less for #-ed, but irregardless of what are the preferences it shouldn't be a mixture. There was no need to call upon Richard :-)

>> I mean original output was "(mod 0x17)", and what's "(mod 0x"

Now it reads
```
# Curve defined by Weierstrass equation
#      y^2 = x^3 + a*x + b
# bignum: 'mod (p)' = 0x17
# bignum: 'a' = 0x1
# bignum: 'b' = 0x1
```
It's better, but I'd say that following would be more appropriate
```
# Curve defined by Weierstrass equation
#      y^2 = x^3 + a*x + b (mod p)
# bignum: 'a' = 0x1
# bignum: 'b' = 0x1
# bignum: 'p' = 0x17
```

dot-asm commented:
BTW, as for removal of empty lines. I rather meant to remove empty lines between strongly related outputs, not all of them. But if I had to choose between empty lines all over and no empty lines, I'd choose the latter [with rationale that INFO: works as kind of delimiter].

dot-asm commented:
Looking at srptest output. Original:
```
# INFO:  @ ../openssl/test/srptest.c:33
# N = EEAF0AB9ADB38DD69C33F80AFA8FC5E86072618775FF3C0B9EA2314C9C256576D674DF7496EA81D3383B4813D692C6E0E0D5D8E250B98BE48E495C1D6089DAD15DC7D7B46154D6B6CE8EF4AD69B15D4982559B297BCF1885C529F566660E57EC68EDBC3C05726CC02FD4CBF4976EAA9AFD5138FE8376435B9FC61D2FC0EB06E3

# INFO:  @ ../openssl/test/srptest.c:33
# g = 2
```
Suggested:
```
# bignum: 'N'
#                                                               bit position
#  eeaf0ab9adb38dd6 9c33f80afa8fc5e8 6072618775ff3c0b 9ea2314c9c256576:  768
#  d674df7496ea81d3 383b4813d692c6e0 e0d5d8e250b98be4 8e495c1d6089dad1:  512
#  5dc7d7b46154d6b6 ce8ef4ad69b15d49 82559b297bcf1885 c529f566660e57ec:  256
#  68edbc3c05726cc0 2fd4cbf4976eaa9a fd5138fe8376435b 9fc61d2fc0eb06e3:    0
# bignum: 'g' = 0x2
```
It's both better and worse. What's worse is that there is no reference to source code. Well, not that original reference to source code was actually very informative, but why not improve output in the process. Suggestion is to add INFO lines *elsewhere*, for example it would seem to be appropriate to throw in some into run_srp_tests() and run_srp_kat().

dot-asm commented on test/test_test.c:
```diff
@@ -480,6 +480,38 @@ static int test_single_eval(void)
+    "-12345678",
+    "1234567890123456789012345678901234567890123456789012"
+        "1234567890123456789012345678901234567890123456789013"
+        "987657"
```
Ew! So point is that you don't trust reader to be able to follow commas, right? And intention ought to be to "startle" reader with jagged formatting. Why not with parenthesis? E.g.
```
    ("1234567890123456789012345678901234567890123456789012"
     "1234567890123456789012345678901234567890123456789013"
     "987657")
```

dot-asm replied:
Just to be clear. This is kind of a side note. What's essential is output improvement.

paulidale replied:
I've removed the indenting.

And no, I don't generally trust people to read the code properly :)  I know how many times I've misread sections...

paulidale commented:
I think I've addressed the main points.

I could add a blank line _before_ TEST_info and TEST_error.  That might help visually break the output.

dot-asm commented:
I'm ready to plus-one this, but it's appropriate to squash some of commits. Could you do that or should I upon commit?

BTW, I'm kind of opposed to formulating it as "XXXXX are the hex digits in *big endian order*". Endianness is about how [multi-byte] values are stored in memory, not how they are printed on screen for human to see. Most notaly, if you attempt to print in "little endian order", would you actually flip hex-digit-wise? No, you'd flip byte-wise [because that's what endianness is about, *byte* order]. Which tells you that "little-endian printout" is utterly counter-intuitive and essentially an unnatural option. *Digits*, be it binary, hex, octal or decimal, are printed in order of descending position.

As for blank lines before TEST. Well, leave it as it is. As already mentioned, INFO: by itself can be viewed as delimiter.

Oh! On side note! I.e. *no* action is expected! I find it annoying that if I want to pipe output to 'more', I have to remember to forward stderr to stdout. I mean I can't remember it, and don't see reason for why I should (because you know, computers for people, not vice versa:-), and it's annoying that I have to re-run it all the time...

paulidale commented:
I'll squash and reduce the commit messages.

The stdout/stderr output question can be considered later.  If the tests are in control of all of the output, I don't see a compelling reason why stdout couldn't be used instead of stderr.

paulidale commented:
Done.

richsalz commented:
commit 3791646.  thanks.
